### DISCUSSION TITLE
Best way to deploy Postal behind a firewall

### DISCUSSION THREAD

**Discussion by Ascar699** [postalserver/postal#1785] [Help with using Postal] [answered]
Makers of Postal, thank you for excellent product.
Please help me with advice on the best deployment approaches.  I have so far achieved the following results:

1. I was able to deploy Postal on a virtual machine connected directly to the internet, so the IP address on the network interface is public and I use Caddy.  This instance of Postal is 100% functional with tracking working properly.  All configuration files are not modified - postal.yml and caddyfile point to all services running on local host address (127.0.0.1).  Everything is good except the fact that I don't have peace of mind about this VM being directly exposed to the internet.

2. I deployed Postal in the same VM (using snapshots) behind HAProxy implemented in pFsense firewall.  Initially:

a) I was getting error 503 - No server available until I changed the bind address from 127.0.0.1 to the private IP address of the VM and HTTP check method from OPTIONS to GET in the backend settings.

b) I could make it only to the login screen and past that I was getting "The change you wanted was rejected" error page.  This got fixed by checking off the "Use forwardfor option" in the Postal frontend in HAProxy.  So, now I can do everything to manage Postal - create orgs, add mail servers, domains.  Mail flows smoothly.

![image](https://user-images.githubusercontent.com/28905723/147476202-907214dc-60d7-43a2-8680-95c360027e0f.png)

To be brief - when my Postal VM is behind HAProxy tracking does not work.  The tracking URLs are opening with the error and no forwarding to the original link happens.  I tried all combinations of changes - changed IP address from localhost to private IP address, protocols from HTTPS to HTTP, ports.
 
![image](https://user-images.githubusercontent.com/28905723/147476256-38f5b06b-fa8d-4153-940f-827db443966f.png)

Please suggest where I should look and what else to try.  Tracking seems to be the only remaining thing that does not work when Postal is behind  the firewall.

**Comment by willpower232** [comment, marked as answer]
If you're seeing that error on tracking, it means you're missing the header mentioned in the documentation https://github.com/postalserver/postal/discussions/1785

> The web server must add the `X-Postal-Track-Host: 1` header so the Postal web server knows to treat requests as tracking requests.

**Reply by Ascar699** [reply]
Thank you for replying.  It was not easy, but I finally figured that the header should be added on the backend side and there should not be the (:) after X-Postal-Track-Host header.

Well, you may congratulate me on Postal finally running behind pFsense!
